compiler/src/iree/compiler/Preprocessing/Common/ApplyPDLPatterns.cpp

```diff
@@ -0,0 +1,515 @@
+
+  // Method to process tensor types.
+  auto processTensorType = [&](RankedTensorType tensorType) {
+    entryPointInputTypes.push_back(bindingType);
+    totalNumDynamicDims += tensorType.getNumDynamicDims();
+  };
+  // Method to process input types.
+  auto processInputType = [&](Type type) {
+    if (type.isIntOrFloat()) {
+      entryPointScalarInputTypes.push_back(type);
+      return;
+    }
+    auto tensorType = type.cast<RankedTensorType>();
+    processTensorType(tensorType);
+  };
+
+  for (auto type : inputTypes) {
+    processInputType(type);
+  }
+  for (auto type : resultTypes) {
+    processTensorType(type.cast<RankedTensorType>());
+  }
+  for (auto type : otherOperandTypes) {
+    processInputType(type);
+  }
+
+  int64_t numTensorOperands = (int64_t)entryPointInputTypes.size();
+  int64_t numScalarOperands = (int64_t)entryPointScalarInputTypes.size();
+  entryPointInputTypes.append(entryPointScalarInputTypes);
+  entryPointInputTypes.append(totalNumDynamicDims, rewriter.getIndexType());
+
+  auto entryPointFnType = FunctionType::get(context, entryPointInputTypes,
+                                            /*results=*/TypeRange{});
+  auto entryPointFn =
+      rewriter.create<func::FuncOp>(loc, entryPointFnName, entryPointFnType);
+  Region &body = entryPointFn.getBody();
+  SmallVector<Location> locs(entryPointInputTypes.size(), loc);
+  rewriter.createBlock(&body, body.begin(), entryPointInputTypes, locs);
+
+  auto entryPointArgs = entryPointFn.getArguments();
+  auto tensorArgs = entryPointArgs.take_front(numTensorOperands);
+  auto scalarArgs = entryPointArgs.slice(numTensorOperands, numScalarOperands);
+  auto dynamicDimArgs = entryPointArgs.take_back(totalNumDynamicDims);
+  SmallVector<Value> callOperands;
+  Value zero = rewriter.create<arith::ConstantIndexOp>(loc, 0);
+
+  // Method to marshal tensor types into call operands.
+  auto marshalTensorTypes = [&](RankedTensorType tensorType) {
+    int64_t numDynamicDims = tensorType.getNumDynamicDims();
+    auto dynamicDims = dynamicDimArgs.take_front(numDynamicDims);
+    auto [basePtr, offset] = getBasePtrAndOffsetForTensor(
+        rewriter, loc, tensorType, tensorArgs.front(), zero, dynamicDims);
+    callOperands.push_back(basePtr);
+    callOperands.push_back(offset);
+    tensorArgs = tensorArgs.drop_front();
+    dynamicDimArgs = dynamicDimArgs.drop_front(numDynamicDims);
+  };
+  // Method to marshal input types into call operands.
+  auto marshalInputTypes = [&](Type type) {
+    if (type.isIntOrFloat()) {
+      callOperands.push_back(scalarArgs.front());
+      scalarArgs = scalarArgs.drop_front();
+      return;
+    }
+    marshalTensorTypes(type.cast<RankedTensorType>());
+  };
+
+  for (auto type : inputTypes) {
+    marshalInputTypes(type);
+  }
+  for (auto type : resultTypes) {
+    marshalTensorTypes(type.cast<RankedTensorType>());
+  }
+  for (auto type : otherOperandTypes) {
+    marshalInputTypes(type);
+  }
```

Comment:
suggestion: can use `llvm::for_each` here and when processing input types https://llvm.org/doxygen/namespacellvm.html#a0b2a153b655ed78a07468297eb4c6256